huaweicloud/services/iam/resource_huaweicloud_identity_provider.go

```diff
@@ -0,0 +1,651 @@
+				ForceNew:     true,
+				Default:      "virtual_user_sso",
+				ValidateFunc: validation.StringInSlice([]string{"virtual_user_sso", "iam_user_sso"}, false),
+			},
+			"access_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringInSlice([]string{"program", "program_console"}, false),
+			},
+			"provider_url": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"client_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"authorization_endpoint": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"scopes": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 10,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"response_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "id_token",
+			},
+			"response_mode": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      "form_post",
+				ValidateFunc: validation.StringInSlice([]string{"fragment", "form_post"}, false),
+			},
+			"signing_key": {
+				Type:     schema.TypeString,
+				Optional: true,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					equal, _ := utils.CompareJsonTemplateAreEquivalent(old, new)
+					return equal
+				},
+			},
+			"metadata": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"login_link": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceIdentityProviderCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conf := meta.(*config.Config)
+	client, err := conf.IAMNoVersionClient(conf.GetRegion(d))
+	if err != nil {
+		return fmtp.DiagErrorf("Error creating HuaweiCloud IAM without version client: %s", err)
+	}
+	// Create a SAML protocol provider.
+	opts := providers.CreateProviderOpts{
+		Enabled:     getEnabledVal(d),
+		SsoType:     d.Get("sso_type").(string),
+		Description: d.Get("description").(string),
+	}
+	name := d.Get("name").(string)
+	logp.Printf("[DEBUG] Create identity options %s : %#v", name, opts)
+	provider, err := providers.Create(client, name, opts)
+	if err != nil {
+		logp.Printf("[ERROR] Failed to create identity provider: %s", err)
+		return diag.FromErr(err)
+	}
+	d.SetId(provider.ID)
+
+	protocol := d.Get("protocol").(string)
+	// Create protocol and mapping
+	err = createMappingAndProtocol(client, d)
+	// Create the provider access type, it only worked on the OIDC protocol.
+	if err == nil && protocol == protocolOIDC {
+		err = createAccessType(conf, d)
+	}
+	// If create mapping or protocol or access type fails, delete the provider.
+	if err != nil {
+		logp.Printf("[ERROR] Failed to create mapping or protocol or access type, delete the provider: %s,", err)
+		resourceIdentityProviderDelete(ctx, d, meta)
+		return diag.FromErr(err)
```

Comment:
if we encounter errors after setting the ID, terraform will mark the resource be tainted, then the resource will be rebuild in the next plan command.  we do not need to delete the idp manually.